Bukkit/src/main/java/com/plotsquared/bukkit/listener/PaperListener.java

```diff
@@ -407,4 +409,27 @@ public void onAsyncTabCompletion(final AsyncTabCompleteEvent event) {
         }
     }
 
+    @EventHandler
+    public void onBeaconEffect(final BeaconEffectEvent event) {
+        Block block = event.getBlock();
+        Location beaconLocation = BukkitUtil.adapt(block.getLocation());
+        Plot beaconPlot = beaconLocation.getPlot();
+        if (beaconPlot == null) {
+            return;
+        }
+
+        Player player = event.getPlayer();
+        Location playerLocation = BukkitUtil.adapt(player.getLocation());
+
+        PlotPlayer<Player> plotPlayer = BukkitUtil.adapt(player);
+        Plot playerStandingPlot = playerLocation.getPlot();
+        if (playerStandingPlot == null || beaconPlot.equals(playerStandingPlot)) {
```

Comment:
It's probably ideal to check if the player is in a plot area, and thus beacon effects can be prevented from "leaking" into the road as well as adjacent plots. Having it as a road flag would also be useful, so checking the road flag container

Reply:
I added a commit. Hope it fixed everything you mentioned